lib/pdk/config/json_schema_namespace.rb

```diff
@@ -0,0 +1,123 @@
+require 'pdk/config/namespace'
+
+module PDK
+  class Config
+    class JSONSchemaNamespace < Namespace
+      # Initialises the PDK::Config::JSONSchemaNamespace object.
+      #
+      # @see PDK::Config::Namespace.initialize
+      #
+      # @option params [String] :schema_file Path to the JSON Schema document
+      def initialize(name = nil, file: nil, parent: nil, persistent_defaults: false, schema_file: nil, &block)
+        super(name, file: file, parent: parent, persistent_defaults: persistent_defaults, &block)
+        @schema_file = schema_file
+      end
+
+      # TODO: What if the document is not valid?!?!? We can never save then
+
+      # The JSON Schema for the namespace
+      #
+      # @return [Hash]
+      def schema
+        document_schema.schema
+      end
+
+      # Whether the schema is valid but empty.
+      #
+      # @return [Boolean]
+      def empty_schema?
+        document_schema.schema.empty?
+      end
+
+      # Name of all the top level properties for the schema
+      #
+      # @return [String[]]
+      def schema_property_names
+        return [] if schema['properties'].nil?
+        schema['properties'].keys
+      end
+
+      # Extends the to_h namespace method to include unmanaged settings
+      #
+      # @see PDK::Config::Namespace.to_h
+      def to_h
+        # This may seem counter-intuitive but we need to call super first as the settings
+        # may not have been loaded yet, which means @unmanaged_settings will be empty.
+        # We call super first to force any file loading and then merge the unmanaged settings
+        settings_hash = super
+        @unmanaged_settings = {} if @unmanaged_settings.nil?
+        @unmanaged_settings.merge(settings_hash)
+      end
+
+      # Validates a document (Hash table) against the schema
+      #
+      # @return [Boolean]
+      def validate_document!(document)
+        ::JSON::Validator.validate!(schema, document)
+      end
+
+      # Sets the list of unmanaged settings. For subclass use only
+      #
+      # @param value [Hash<String, Object]] A hashtable of all unmanaged settings which will be persisted, but not visible
+      #
+      # @private
+      def assign_unmanaged_settings(value)
+        @unmanaged_settings = value
+      end
```

Comment:
Should this be `protected` if it's only for use by subclasses?

Reply:
TIL `protected`.  I made it public because I use it in tests.  Could be a code smell that the code is written as testable as it could.

Reply:
If you just need to call it for test setup you can always bypass the visibility check with https://ruby-doc.org/core-2.6.4/Object.html#method-i-send

Or you could use https://ruby-doc.org/core-2.6.4/Object.html#method-i-instance_variable_set to bypass the method entirely if it doesn't do anything special.

Also should we just call this method `def unmanaged_settings=(value)` ?

Reply:
Went for a protected `attr_writer` for unmanaged_settings.

Added a TODO that we should probably subclass instead of using `instance_variable_set` but that shouldn't stop merging.